baseapp/baseapp.go

```diff
@@ -653,9 +661,22 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		anteEvents = events.ToABCIEvents()
 	}
 
+	if mode == runTxModeCheck {
+		err = app.mempool.Insert(ctx, tx.(mempool.Tx))
+		if err != nil {
+			return gInfo, nil, anteEvents, priority, err
+		}
+	} else if mode == runTxModeDeliver {
+		err = app.mempool.Remove(tx.(mempool.Tx))
+		if err != nil && !errors.Is(err, mempool.ErrTxNotFound) {
+			return gInfo, nil, anteEvents, priority,
+				fmt.Errorf("failed to remove tx from mempool: %w", err)
+		}
+	}
```

Comment:
How about `runTxModeReCheck`? My understanding is that the SDK will have to return valid/invalid depending on whether the Tx is _still_ in `app.mempool` or not.

Reply:
I didn't think about ReCheckTx. What should happen? If the tx is not in the mempool, we error? If it is in the mempool, we just do nothing regarding the mempool?

Reply:
Tendermint calls re-CheckTx after delivering every block at Commit time, to make sure that the transactions in its mempool are still valid when applied to the new App state.
With that, and your use case in mind:
* If `CheckTx` is called **without** the _re-checkTx_ flag then
  * if the Tx is **not** in the SDK's mempool: validate it, if valid, add it to the SDK's mempool, and return "valid" (normal case)
  * if the Tx is already in the SDK's mempool: something bad happened. My advice: log an error and return "valid" (if you don't return valid, the transaction will stop being gossiped, whereas it _is_ in the SDK's mempool)
* If `CheckTx` is called **with** the _re-checkTx_ flag then
  * if the Tx is in the SDK's mempool: do nothing and return "valid" (normal case)
  * if the Tx is **not** in the SDK's mempool: something bad happened. The Tx can't be in the recently delivered block, as Tendermint removes those from its mempool before calling re-CheckTx; and any transaction newly received via gossip is blocked because Tendermint locks its mempool during Commit (see the [spec](https://github.com/tendermint/tendermint/blob/v0.37.x/spec/abci/abci%2B%2B_app_requirements.md#commit)). My advice: log an error, then validate the Tx and, if valid, add it to the mempool 

HIH

Reply:
ABCI 1.0 integration doesn't change CheckTx semantics in the SDK. I mainly want to know what we do re _app-side mempool_ -- we don't care really about Tendermint's mempool.

Looks like during Re-CheckTx, if the tx is NOT in the app mempool, we just log and ignore 👍